### PR TITLE
Http request

### PR DESCRIPTION
**Pull request by ShayanDeSh**
This is a simple implemention for creating HTTP packets and writing them directly to a file descriptor. #37

### CODE REVIEW

File: http_request.c [hunk 1]
```diff
@@ -0,0 +1,64 @@
+#include "http_request.h"
+#include <unistd.h>
+#include <stdio.h>
+
+int write_verb_path(const int fd, const char* verb, const char* path) {
```
Comment (pylover) [inline http_request.c]: `const int fd` Nice idea. :+1:

File: http_request.c [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    return dprintf(fd, "%s %s %s\r\n", verb, path, HTTP_VERSION);
+}
+
+int write_host(const int fd, const char* host) {
+    return dprintf(fd, "HOST: %s\r\n", host);
+}
+
+int write_headers(const int fd, char* headers[], const int header_count) {
+    int i, err, sum = 0;
```
Comment (pylover) [inline http_request.c]: Please define at most one variable per line.

```C
int foo;
int bar;
int baz;
```

File: http_request.c [hunk 3]
```diff
@@ -0,0 +1,64 @@
+#include "http_request.h"
+#include <unistd.h>
+#include <stdio.h>
+
+int write_verb_path(const int fd, const char* verb, const char* path) {
+    return dprintf(fd, "%s %s %s\r\n", verb, path, HTTP_VERSION);
+}
+
+int write_host(const int fd, const char* host) {
+    return dprintf(fd, "HOST: %s\r\n", host);
```
Comment (pylover) [inline http_request.c]:
```C
return dprintf(fd, "HOST: %s" RN, host);
```

File: http_request.c [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+int write_headers(const int fd, char* headers[], const int header_count) {
+    int i, err, sum = 0;
+    for (i = 0; i < header_count; i++) {
+        err = dprintf(fd, "%s\r\n", headers[i]);
+        if (err < 0)
```
Comment (pylover) [inline http_request.c]: Please always use `{}` for if blocks.

```C
if (err) {
    // handle err
}
```

File: http_request.c [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+            return -1;
```
Comment (pylover) [inline http_request.c]:
```C
#include <common.h>
.....
return ERR;
```

File: http_request.c [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    return sum;
+}
+
+int write_body(const int fd, const char* body) {
+    int err, sum = 0;
+    err = dprintf(fd, "Content-Length: %ld\r\n\r\n", strlen(body));
+    if (err < 0)
+        return -1;
+    sum += err;
+    err = write(fd, body, strlen(body));
```
Comment (pylover) [inline http_request.c]: Do not use `strlen()` here to get rid of possible null characters inside body. 

And please pass the length of `body` from the caller.

File: http_request.c [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    if (err < 0)
+        return -1;
+    return sum + err;
+}
+
+int send_request(
```
Comment (pylover) [inline http_request.c]: Please do not folow the rule: `one-item-per-line` jsut for the function definitions.

an example of function definition:

```C
int rb_dryread_until(struct ringbuffer *b, char *data, size_t len,
                     char *delimiter, size_t dlen, size_t *readlen);
```

File: http_request.h [hunk 8]
```diff
@@ -0,0 +1,33 @@
+#ifndef HTTP_REQUEST_H
+#define HTTP_REQUEST_H
+
+#include <string.h>
+
+typedef void (*callback)(int status, char* body, void* arg);
+
+#define HTTP_VERSION "HTTP/1.0"
+
+#define CR "\r\n"
```
Comment (pylover) [inline http_request.h]: The `RN` macro is already defined inside `common.h`

File: http_request.h [hunk 9]
```diff
@@ -0,0 +1,33 @@
+#ifndef HTTP_REQUEST_H
+#define HTTP_REQUEST_H
+
+#include <string.h>
+
+typedef void (*callback)(int status, char* body, void* arg);
+
+#define HTTP_VERSION "HTTP/1.0"
```
Comment (pylover) [inline http_request.h]: We need both `1.0` and `1.1` HTTP versions.

File: tests/test_http_request.c [hunk 10]
```diff
@@ -0,0 +1,106 @@
+#include "fixtures/assert.h"
+#include "logging.h"
+#include "http_request.h"
+#include <sys/types.h>
+#include <fcntl.h>
+#include <unistd.h>
+
+#define PATH "/tmp/http_request.txt"
```
Comment (pylover) [inline tests/test_http_request.c]: Use `tempnam(3)` instead of the static temporary file name.

File: tests/test_http_request.c [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+
+int
```
Comment (pylover) [inline tests/test_http_request.c]: Use macro instead of function.

File: tests/test_http_request.c [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+    read(fd, response, len);
+    eqstr(vpv, response);
+    free(response);
+    clean_fd();
+}
+
+void
+test_write_host() {
+    int fd, len;
+    char* host = "google.com";
```
Comment (pylover) [inline tests/test_http_request.c]: Bad idea:

```C
char* c;
char * c;
```

Good idea:

```C
char *c;
```

File: tests/test_http_request.c [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+    char *response, *expected;
+    fd = mock_fd();
+    len = write_host(fd, host);
+    notequalint(-1, len);
+    expected = (char*) malloc(len * sizeof(char));
```
Comment (pylover) [inline tests/test_http_request.c]: Do not allocate here.

Simply use: 

```C
eqstr(" ....", ...)
```

With static host argument.

File: tests/test_http_request.c [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+    sprintf(expected, "HOST: %s\r\n", host);
+    response = (char*) malloc(len * sizeof(char));
```
Comment (pylover) [inline tests/test_http_request.c:41]: And always free what you allocated.

File: tests/test_http_request.c [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+    eqstr(expected, response);
+    clean_fd();
+}
+
+void
+test_write_body() {
+    int fd, len;
+    char* response;
+    char* expected;
+    char* body = "{\"name\": \"Dennis\", \"lastname\": \"Ritchie\"}";
```
Comment (pylover) [inline tests/test_http_request.c]: Simply use `foo bar baz` instead of ugly json `{}` and unwanted quotes: `"`.

File: tests/test_http_request.c [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+    fd = mock_fd();
+    len = write_body(fd, body);
+    notequalint(-1, len);
```
Comment (pylover) [inline tests/test_http_request.c]:
```C
notequalint(ERR, len)
```

File: tests/test_http_request.c [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+    expected = (char*) malloc(len * sizeof(char));
+    sprintf(expected, "Content-Length: %ld\r\n\r\n%s", strlen(body), body);
+    read(fd, response, len);
+    eqstr(expected, response);
+    clean_fd();
+}
+
+int
+main() {
+    test_write_verb_path();
```
Comment (pylover) [inline tests/test_http_request.c]: Please insert:

```C
log_setlevel(LL_DEBUG)
```

At the first line of the test `main()`  functions.

**Review by pylover — CHANGES_REQUESTED** [review]
First, thanks a lot for this PR.

Please pass tests before sending any PR.

**Comment by ShayanDeSh** [comment]
As I noticed from action logs I think the reason pipeline fails is because of coveralls token. And there is a test for test_server_cli which fails and I noticed that one failed on master branch pipline too but the check is given to the branch.

And I also fixed the coding style you mentioned.

**Comment by pylover** [comment]
Let I check it, coveralls token is already available in repo secrets.

And if any test is failing in your machine, please submit an issue and provide appropriate information to investigate and fix it.

**Comment by ShayanDeSh** [comment]
I added a condition to pipeline so coveralls would be skipped on PRs #43

**Comment by pylover** [comment]
I'll merge it asap.

**Pull request merged**